### PR TITLE
NLL: Suggest extraction of subcomputation into a new binding

### LINKED ISSUE #47349: Improve diagnostics to guide users past Borrow checking regression in brotli
I mean regression in the sense that code that used to compile doesn’t anymore. Maybe that is an intentional soundness fix.

Compiling https://crates.io/crates/brotli / https://github.com/dropbox/rust-brotli with:

* rustc 1.25.0-nightly (61452e506 2018-01-09): OK
* rustc 1.25.0-nightly (f62f77403 2018-01-10): borrowck error:

```rust
error[E0502]: cannot borrow `self.buckets_` as immutable because it is also borrowed as mutable
   --> src/enc/backward_references.rs:339:51
    |
338 |     (*self).buckets_.slice_mut()[(key as (usize)).wrapping_add((cur_ix >> 3)
    |     ---------------- mutable borrow occurs here
339 |                                     .wrapping_rem(self.buckets_.BUCKET_SWEEP() as
    |                                                   ^^^^^^^^^^^^^ immutable borrow occurs here
340 |                                                   usize))] = cur_ix as (u32);
    |                                                          - mutable borrow ends here

error: aborting due to previous error

error: Could not compile `brotli`.
```

Commit range: https://github.com/rust-lang/rust/compare/61452e506...f62f77403. https://github.com/rust-lang/rust/pull/47167 seems relevant. CC @ivanbakel @eddyb

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #47349.

r? @nikomatsakis

### CODE REVIEW

File: src/test/ui/borrowck/borrowck-suggest-extraction-of-subcomputation.nll.stderr [hunk 1]
```diff
@@ -0,0 +1,19 @@
+error[E0502]: cannot borrow `buckets` as immutable because it is also borrowed as mutable
+  --> $DIR/borrowck-suggest-extraction-of-subcomputation.rs:25:70
+   |
+LL |     buckets.slice_mut()[(key as usize).wrapping_add(22).wrapping_rem(buckets.sweep())] = 22;
+   |     -----------------------------------------------------------------^^^^^^^----------
```
Comment (KiChjang) [inline src/test/ui/borrowck/borrowck-suggest-extraction-of-subcomputation.nll.stderr]: I wonder if there's an issue filed for this diagnostic change? It looks to be a worse version of what we have in the AST borrowck.

**Comment by KiChjang** [comment]
The complete MIR for the `foo` function in the test file is:

```rust
// MIR for `foo`
// source = MirSource { def_id: DefId(0/0:9 ~ borrowck_suggest_extraction_of_subcomputation[317d]::foo[0]), promoted: None }
// pass_name = SimplifyCfg-initial
// disambiguator = after

fn foo(_1: B) -> (){
    let mut _0: ();                      // return place
    scope 1 {
    }
    scope 2 {
        let _2: u32;                     // "key" in scope 2 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:24:9: 24:12
    }
    let mut _3: &mut [u32];
    let mut _4: &mut B;
    let mut _5: usize;
    let mut _6: usize;
    let mut _7: usize;
    let mut _8: u32;
    let mut _9: usize;
    let mut _10: &B;
    let mut _11: usize;
    let mut _12: bool;

    bb0: {                              
        StorageLive(_2);                 // bb0[0]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:24:9: 24:12
        _2 = const 0u32;                 // bb0[1]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:24:15: 24:19
                                         // ty::Const
                                         // + ty: u32
                                         // + val: Scalar(Bits { size: 4, bits: 0 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:24:15: 24:19
                                         // + ty: u32
                                         // + literal: Const { ty: u32, val: Scalar(Bits { size: 4, bits: 0 }) }
        StorageLive(_3);                 // bb0[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:24
        StorageLive(_4);                 // bb0[3]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:12
        _4 = &mut _1;                    // bb0[4]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:12
        _3 = const SliceWrapperMut::slice_mut(move _4) -> [return: bb2, unwind: bb3]; // bb0[5]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:24
                                         // ty::Const
                                         // + ty: for<'r> fn(&'r mut B) -> &'r mut [u32] {<B as SliceWrapperMut<u32>>::slice_mut}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:24
                                         // + ty: for<'r> fn(&'r mut B) -> &'r mut [u32] {<B as SliceWrapperMut<u32>>::slice_mut}
                                         // + literal: Const { ty: for<'r> fn(&'r mut B) -> &'r mut [u32] {<B as SliceWrapperMut<u32>>::slice_mut}, val: Scalar(Bits { size: 0, bits: 0 }) }
    }

    bb1: {                               // cleanup
        resume;                          // bb1[0]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:23:1: 27:2
    }

    bb2: {                              
        StorageDead(_4);                 // bb2[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:23: 25:24
        StorageLive(_5);                 // bb2[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:86
        StorageLive(_6);                 // bb2[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:56
        StorageLive(_7);                 // bb2[3]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:39
        StorageLive(_8);                 // bb2[4]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:26: 25:29
        _8 = _2;                         // bb2[5]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:26: 25:29
        _7 = move _8 as usize (Misc);    // bb2[6]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:39
        StorageDead(_8);                 // bb2[7]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:38: 25:39
        _6 = const core::num::<impl usize>::wrapping_add(move _7, const 22usize) -> [return: bb4, unwind: bb3]; // bb2[8]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:56
                                         // ty::Const
                                         // + ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_add}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:56
                                         // + ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_add}
                                         // + literal: Const { ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_add}, val: Scalar(Bits { size: 0, bits: 0 }) }
                                         // ty::Const
                                         // + ty: usize
                                         // + val: Scalar(Bits { size: 8, bits: 22 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:53: 25:55
                                         // + ty: usize
                                         // + literal: Const { ty: usize, val: Scalar(Bits { size: 8, bits: 22 }) }
    }

    bb3: {                               // cleanup
        EndRegion();                     // bb3[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:87
        drop(_1) -> bb1;                 // bb3[1]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:27:1: 27:2
    }

    bb4: {                              
        StorageDead(_7);                 // bb4[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:55: 25:56
        StorageLive(_9);                 // bb4[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
        StorageLive(_10);                // bb4[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:77
        _10 = &_1;                       // bb4[3]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:77
        _9 = const Sweep::sweep(move _10) -> [return: bb5, unwind: bb6]; // bb4[4]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
                                         // ty::Const
                                         // + ty: for<'r> fn(&'r B) -> usize {<B as Sweep>::sweep}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
                                         // + ty: for<'r> fn(&'r B) -> usize {<B as Sweep>::sweep}
                                         // + literal: Const { ty: for<'r> fn(&'r B) -> usize {<B as Sweep>::sweep}, val: Scalar(Bits { size: 0, bits: 0 }) }
    }

    bb5: {                              
        EndRegion();                     // bb5[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
        StorageDead(_10);                // bb5[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:84: 25:85
        _5 = const core::num::<impl usize>::wrapping_rem(move _6, move _9) -> [return: bb7, unwind: bb3]; // bb5[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:86
                                         // ty::Const
                                         // + ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_rem}
                                         // + val: Scalar(Bits { size: 0, bits: 0 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:25: 25:86
                                         // + ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_rem}
                                         // + literal: Const { ty: fn(usize, usize) -> usize {core::num::<impl usize>::wrapping_rem}, val: Scalar(Bits { size: 0, bits: 0 }) }
    }

    bb6: {                               // cleanup
        EndRegion();                     // bb6[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
        goto -> bb3;                     // bb6[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:70: 25:85
    }

    bb7: {                              
        StorageDead(_9);                 // bb7[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:85: 25:86
        StorageDead(_6);                 // bb7[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:85: 25:86
        _11 = Len((*_3));                // bb7[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:87
        _12 = Lt(_5, _11);               // bb7[3]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:87
        assert(move _12, "index out of bounds: the len is move _11 but the index is _5") -> [success: bb8, unwind: bb3]; // bb7[4]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:87
    }

    bb8: {                              
        EndRegion();                     // bb8[0]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:87
        (*_3)[_5] = const 22u32;         // bb8[1]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:5: 25:92
                                         // ty::Const
                                         // + ty: u32
                                         // + val: Scalar(Bits { size: 4, bits: 22 })
                                         // mir::Constant
                                         // + span: .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:90: 25:92
                                         // + ty: u32
                                         // + literal: Const { ty: u32, val: Scalar(Bits { size: 4, bits: 22 }) }
        StorageDead(_3);                 // bb8[2]: scope 1 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:25:92: 25:93
        _0 = ();                         // bb8[3]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:23:77: 27:2
        StorageDead(_2);                 // bb8[4]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:27:1: 27:2
        drop(_1) -> [return: bb9, unwind: bb1]; // bb8[5]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:27:1: 27:2
    }

    bb9: {                              
        return;                          // bb9[0]: scope 0 at .\src\test\ui\borrowck\borrowck-suggest-extraction-of-subcomputation.rs:27:2: 27:2
    }
}
```

**Comment by KiChjang** [comment]
One of the failed tests is quite interesting:

```
[00:51:17] thread '[ui (nll)] ui/borrowck/mut-borrow-in-loop.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3149:9
[00:51:17] 
[00:51:17] ---- [ui (nll)] ui/codemap_tests/one_line.rs stdout ----
[00:51:17] diff of stderr:
[00:51:17] 
[00:51:17] 7	   |     |      second mutable borrow occurs here
[00:51:17] 8	   |     first mutable borrow occurs here
[00:51:17] 9	   |     borrow later used here
[00:51:17] +	   |
[00:51:17] +	note: consider extracting this into a `let`-binding
[00:51:17] +	  --> $DIR/one_line.rs:13:12
[00:51:17] +	   |
[00:51:17] +	LL |     v.push(v.pop().unwrap()); //~ ERROR cannot borrow
[00:51:17] +	   |            ^^^^^^^
[00:51:17] 10	
[00:51:17] 11	error: aborting due to previous error
```

It does feel like this PR is making the right recommendation for this test, but the rest of the failed test cases indicate that the current check requires more refinement.

File: src/test/ui/borrowck/borrowck-suggest-extraction-of-subcomputation.nll.stderr (hunk 1 above)
Comment (nikomatsakis) [reply src/test/ui/borrowck/borrowck-suggest-extraction-of-subcomputation.nll.stderr]: Are you referring to the "really big span"?

Comment (nikomatsakis) [reply src/test/ui/borrowck/borrowck-suggest-extraction-of-subcomputation.nll.stderr]: I'm not sure if there is an issue but I agree it's a diagnostic regression

**Comment by KiChjang** [comment]
@nikomatsakis To explain more what I was suggesting in Zulip, consider the MIR I've posted above:

In bb0[4], we have an assignment of a mutable reference of _1 to _4, which then immediately gets used as an argument to a function in the next statement, the return value of which is stored in _3. Since there is no other statement between the assignment and the use of _4, there would not be a chance to run the analysis at all. I think a more correct formulation of the rule would be to check when _1 starts being borrowed and when it's StorageDead statement is emitted.

**Comment by nikomatsakis** [comment]
@KiChjang 

I'm a bit confused by your comment. 

> ...since there is no other statement between the assignment and the use of _4, there would not be a chance to run the analysis at all...

I'm not sure what you mean here -- run which analysis? Ah, I guess you are referring to these conditions that [I wrote earlier](https://github.com/rust-lang/rust/issues/47349#issuecomment-410852205):

> I am imagining something like this:
> 
> - If the "starting borrow" is written into a temporary, and that temporary (or some other temporary that it is copied into) winds up being ultimately used as the argument to a call;
> - and the "error case" comes between that starting borrow and the call
> 
> then we might add a note like
> 
> "consider extracting this into a fresh variable with a let"

I agree that what I wrote here doesn't seem exactly correct. When I wrote 'the call', of course, I was referring not to the call to `IndexMut` but rather the call to `wrapping_rem` that occurs later.

One challenge is that this suggestion is sort of a "tree-based" one -- we basically want to see whether we could re-order some subcomputation (like `buckets.sweep()`) so as to make it occur before the borrow that is causing the conflict. 

Specifically I was looking for the case where you have `foo(xxx, ..., yyy)`, where the computation `xxx` is causing the mutable borrow somehow, and the computation `yyy` is causing the (conflicting) shared borrow. 

I don't really think that looking for `StorageDead` is necessarily what we are after.

**Comment by nikomatsakis** [comment]
Sorry, that comment isn't really a "conclusion" -- more just me musing to myself. =) I have to run now though, will try to think a bit more in a bit.

**Comment by nikomatsakis** [comment]
Let me know though if what I wrote makes sense and is 'aligned' with how you are thinking about the problem. =)

**Comment by nikomatsakis** [comment]
@KiChjang so what do you think? After our discussion on Zulip I'm feeling a bit uncertain how to proceed here.

**Comment by KiChjang** [comment]
@nikomatsakis Same here unfortunately, there doesn't seem to be a clear path forward, except for the "availability" analysis you mentioned, which is much more involved than what this PR is and I think that one is worth specc'ing out first. I have a feeling that anything else we do currently would be an inaccurate approximation of what the desired behavior is.

**Comment by nikomatsakis** [comment]
OK, I guess we'll close this PR for now. We should raise the original issue for discussion today perhaps.

**Pull request closed without merging**